src/cnaas_nms/tools/security.py

```diff
@@ -14,10 +28,124 @@ def jwt_required(fn):
     else:
         return fn
 
-
 def get_jwt_identity():
     """
     This function overides the identity when needed.
-
     """
     return get_jwt_identity_orig() if api_settings.JWT_ENABLED else "admin"
+
+logger = get_logger()
+oauth_required = ResourceProtector()
+
+
+
+
+class MyBearerTokenValidator(BearerTokenValidator):
+    keys: Mapping[str, Any] = {}
+
+    def authenticate_token(self, token_string: str):
+        """Check if token is active.
+
+        If JWT is disabled, we return because no token is needed.
+
+        We decode the header and check if it's good.
+
+        We decode the token using the keys.
+        We first check if we can decode it, if not we request the keys.
+        The decode function also checks if it's not expired.
+        We get de decoded _token back, but for now we do nothing with this.
+
+        Input
+            token_string(str): The tokenstring
+        Returns:
+            token(dict): Dictionary with access_token, decoded_token, token_type, audience, expires_at
+
+        """
+        # If JWT is disabled, no token is needed
+        if not api_settings.JWT_ENABLED:
+            return "no-token-needed"
+        # First decode the header
+        try:
+            unverified_header = jwt.get_unverified_header(token_string)
+        except exceptions.JWSError as e:
+            raise InvalidTokenError(e)
+        except exceptions.JWTError as e:
+            raise InvalidTokenError(e)
+
+        # decode the token
+        algorithm = unverified_header.get('alg')
+        try:
+            decoded_token = jwt.decode(token_string, self.keys, algorithms=algorithm, audience=auth_settings.OIDC_CLIENT_ID)
+        except exceptions.ExpiredSignatureError as e:
+            raise ExpiredSignatureError(e)
+        except exceptions.JWTClaimsError as e:
+            try:
+                # with this exception, the call was too fast or the computer too slow
+                logger.info('Checked token too fast (or computer is too slow), retry')
+                sleep(1)
+                decoded_token = jwt.decode(token_string, self.keys, algorithms=algorithm, audience=auth_settings.OIDC_CLIENT_ID)
+            except exceptions.ExpiredSignatureError as e:
+                raise ExpiredSignatureError(e)
+            except exceptions.JWKError as e:
+                logger.error("Invalid Key")
+                raise InvalidKeyError(e)
+        except exceptions.JWKError:
+            try:
+                # with this exception, we first try to reload the keys
+                logger.info('JWT.decode didnt work. Get the keys and retry.')
+                response = requests.get(url=auth_settings.OIDC_CONF_WELL_KNOWN_URL.split('.well-known')[0] + 'oidc/certs')
```

Comment:
Same as below, should be dynamic

Reply:
Thanks, changed it to follow your recommendation.